pymc/distributions/multivariate.py

```diff
@@ -2380,3 +2383,169 @@ def logp(value, alpha, K):
+    #     if not isinstance(dist, TensorVariable) or not isinstance(
+    #         dist.owner.op, (RandomVariable, SymbolicRandomVariable)
+    #     ):
+    #         raise ValueError(
+    #             f"Censoring dist must be a distribution created via the `.dist()` API, got {type(dist)}"
+    #         )
+    #     if dist.owner.op.ndim_supp > 0:
+    #         raise NotImplementedError(
+    #             "Censoring of multivariate distributions has not been implemented yet"
+    #         )
+    #     check_dist_not_registered(dist)
+    #     return super().dist([dist, lower, upper], **kwargs)
+
+    @classmethod
+    def rv_op(cls, sigma, zerosum_axes, size=None):
+        if size is None:
+            zerosum_axes_ = np.asarray(zerosum_axes)
+            # just a placeholder size to infer minimum shape
+            size = np.ones(
+                max((max(np.abs(zerosum_axes_) - 1), max(zerosum_axes_))) + 1, dtype=int
+            ).tolist()
+
+        # check if zerosum_axes is valid
+        normalize_axis_tuple(zerosum_axes, len(size))
+
+        normal_dist = ignore_logprob(pm.Normal.dist(sigma=sigma, size=size))
+        normal_dist_, sigma_ = normal_dist.type(), sigma.type()
+
+        # Zerosum-normaling is achieved by substracting the mean along the given zerosum_axes
+        zerosum_rv_ = normal_dist_
+        for axis in zerosum_axes:
+            zerosum_rv_ -= zerosum_rv_.mean(axis=axis, keepdims=True)
+
+        return ZeroSumNormalRV(
+            inputs=[normal_dist_, sigma_],
+            outputs=[zerosum_rv_],
+            zerosum_axes=zerosum_axes,
+            ndim_supp=0,
+        )(normal_dist, sigma)
+
+
+@_change_dist_size.register(ZeroSumNormalRV)
+def change_zerosum_size(op, normal_dist, new_size, expand=False):
+    normal_dist, sigma = normal_dist.owner.inputs
+    if expand:
+        new_size = tuple(new_size) + tuple(normal_dist.shape)
+    return ZeroSumNormal.rv_op(sigma=sigma, zerosum_axes=op.zerosum_axes, size=new_size)
+
+
+@_moment.register(ZeroSumNormalRV)
+def zerosumnormal_moment(op, rv, *rv_inputs):
+    return at.zeros_like(rv)
+
+
+@_default_transform.register(ZeroSumNormalRV)
+def zerosum_default_transform(op, rv):
+    return ZeroSumTransform(op.zerosum_axes)
+
+
+@_logprob.register(ZeroSumNormalRV)
+def zerosumnormal_logp(op, values, normal_dist, sigma, **kwargs):
+    (value,) = values
+    shape = value.shape
+    _deg_free_shape = at.inc_subtensor(shape[at.as_tensor_variable(op.zerosum_axes)], -1)
+    _full_size = at.prod(shape)
+    _degrees_of_freedom = at.prod(_deg_free_shape)
+    zerosums = [
+        at.all(at.isclose(at.mean(value, axis=axis), 0, atol=1e-9)) for axis in op.zerosum_axes
+    ]
+    # out = at.sum(
+    #     pm.logp(dist, value) * _degrees_of_freedom / _full_size,
+    #     axis=op.zerosum_axes,
+    # )
+    # figure out how dimensionality should be handled for logp
+    # for now, we assume ZSN is a scalar distribut, which is not correct
```

Comment:
You **do need to sum** across the zero sum axes in the logp. That’s what gives you the formula you are using!
The way in which is logp is derived is by noting that a zerosumNormal is an MvNormal, with a singular covariance matrix. In the old PR about this, I gave a longer explanation on this but the short reason for using the proper logp is that we can use the special MvNormal as a baseline comparison of the logp values

Reply:
Yeah this is still confusing to me too why we don't use the sum here. I deferred to @ricardoV94, who explained to me that, since we're (wrongly) assuming for now that ZSN is a scalar distribution, the results will be correct -- and they indeed seem to be, at least according to the tests we did.
@ricardoV94 you probably wanna chime in?

Reply:
If you sum the logp across the zerosum axis they should only be allowed to be to the rightmost position. Everything that works with multivariate distributions in Aesara, Aeppl and PyMC expects that at the moment.

Also you will need to add an internal variable for the shape dimensions that would correspond to the support dimensions since size cannot include those, like `steps` for timeseries.

Given those constraints I suggested we treat it as a scalar internally temporarily. 

Also because of these API things we have to figure out, I think this makes more sense to go to pymc-experimental first @AlexAndorra

Reply:
That's clearer, thanks @ricardoV94.

> Given those constraints I suggested we treat it as a scalar internally temporarily

And to make sure: this doesn't change anything mathematically?

> because of these API things we have to figure out, I think this makes more sense to go to pymc-experimental first

Good point. We should probably wait for Aesara to relax its constraints on multivariate distributions before adding ZSN to PyMC. This would allow to sum the logp and alleviate the need for an internal variable at the same time.
I wonder if I can just just transfer this PR and its history to pymc-experimental 🤔

Reply:
I honestly don’t mind changing our ZSN design to enforce the zero sum axes to be at the right. Users would be able to transpose the results to the order they care about

Reply:
I might be missing something, but couldn't we treat it as if `ndim_support` is the number of axis to the right of the first zero_sum constraint axis? So something like this:
```
pm.ZeroSumNormal("x", dims=("a", "b", "c"), zerosum_axes=["b"])
```
Would be a RV over `("b", "c")`, ie the elemwise logp would have dim `("a",)`.

Reply:
> I honestly don’t mind changing our ZSN design to enforce the zero sum axes to be at the right

I agree. @ricardoV94 is there any way we enforce this constraint, or it's an assumption we make in the code?

> I might be missing something, but couldn't we treat it as if ndim_support is the number of axis to the right of the first zero_sum constraint axis?

I think it's possible (need @ricardoV94's confirmation). In any case that'd mean the zerosum axes are to the rightmost position

Reply:
> In any case that'd mean the zerosum axes are to the rightmost position

Why? It could still be the first axes, only all following axes would be interpreted as core dims, right?

Reply:
Yeah we could also do that, but I still like the only consecutive zerosum axis +  rightmost more.

Reply:
The rightmost position enforcement is done (see latest updates).
For future reference, when @ricardoV94 says he "likes" a solution, that doesn't mean said solution is _easy_ to implement